### PR TITLE
Initial support for Mesos master

### PR DESCRIPTION
**Pull request by tripledes**
This PR adds support for Mesos masters metrics. Let me know if anything should be change/modified.

Last commit could be reverted, although I think it could help avoiding issues with existing users when adding more mesos components.

### CODE REVIEW

**Comment by tripledes** [comment]
Sorry, missed that issue reported by the test, will fix right away.

File: plugins/inputs/mesos/mesos_test.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+	a := "wrong_key"
+	expect := []string{}
+	got := masterBlocks(a)
+
+	if !reflect.DeepEqual(got, expect) {
+		t.Errorf("Expected empty string slice, got: %v", got)
+	}
+}
+
+func TestSampleConfig(t *testing.T) {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos_test.go]: you don't need to test this

File: plugins/inputs/mesos/mesos_test.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+`
+
+	got := new(Mesos).SampleConfig()
+
+	if expect != got {
+		t.Errorf("Got %s", got)
+	}
+}
+
+func TestDescription(t *testing.T) {
```
Comment (sparrc) [inline plugins/inputs/mesos/mesos_test.go]: don't test this either

**Comment by sparrc** [comment]
looks good, thanks @tripledes, I'll do a full review tomorrow

**Comment by tripledes** [comment]
@sparrc Thanks, I'll remove those tests later today.

**Comment by tripledes** [comment]
Hmm seems the test failed but not due to my changes :)

**Comment by sparrc** [comment]
Code looks good, but can you write a README based on this template?: https://github.com/influxdata/telegraf/blob/master/plugins/inputs/EXAMPLE_README.md

**Comment by tripledes** [comment]
@sparrc Just added the README, test seems to fail because of something else.

**Comment by sparrc** [comment]
@tripledes I pulled down your changes but I'm also getting the config file panic that CircleCI is hitting. I can't see anything in your code that would cause this, still investigating...

**Comment by sparrc** [comment]
@tripledes I'm sorry but I really can't figure out what the problem is here, when I create a sample-config file using your code, I get toml parse panic. When I simply delete the `inputs.mesos` table, I don't get the panic (using your code or the master code)

**Comment by sparrc** [comment]
these are basically the steps to reproduce:

1. `telegraf -sample-config > telegraf.conf`
1. `telegraf -config telegraf.conf -input-filter cpu:mem -test`

**Comment by tripledes** [comment]
@sparrc I'll have a look, thanks for taking the time to check.

**Comment by tripledes** [comment]
@sparrc not sure about this, but feels like a bug in the TOML parser, could you please give a try to the following config snippet?

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms
  timeout = 100
  # A list of Mesos masters. e.g. master1:5050, master2:5050, etc.
  # The port can be skipped if using the default (5050)
  # Default value is localhost:5050.
  masters = ["localhost:5050"]
  # Metrics groups to be collected.
  master_collections = ["resources"]
```

As soon as I add another comment on top of `master_collections` or more elements to it, the parser crashes. There are currently some issues opened against naoina/toml that feel somehow related.

**Comment by tripledes** [comment]
@sparrc quick update, seems related to multi-line comments, leaving the sample config to the following, works:

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # Default value is localhost:5050.
  masters = ["localhost:5050"]
  # Default, all enabled.
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

Please, give a try and let me know whether that would be enough for the sample config and I'll modify the method.

Still feels like something funny is going on in the parser.

Cheers.

**Comment by sparrc** [comment]
hm, definitely seems like a bug in the parser, I don't understand....there are other plugins with multi-line comments, correct? Why does this seem to only affect yours?

**Comment by tripledes** [comment]
Good question :)

So, different combinations:

* The following ones work:
```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # Default value is localhost:5050.
  masters = ["localhost:5050"]
  # Metrics groups to be collected.
  # Default, all enabled.
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # The quick brown fox jumps over the lazy dog
  masters = ["localhost:5050"]
  # The quick brown fox jumps over the lazy dog
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # The quick brown fox jumps over the lazy dog
  masters = ["localhost:5050"]
  # The quick brown fox jumps over the lazy dog
  # blahblah
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # The quick brown fox jumps over the lazy dog
  # blahblah
  masters = ["localhost:5050"]
  # The quick brown fox jumps over the lazy dog
  # blahblah
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

* The following and any other attempt to add more chars to the comments, don't work:

```toml
# Telegraf plugin for gathering metrics from N Mesos masters
[[inputs.mesos]]
  # Timeout, in ms.
  timeout = 100
  # The quick brown fox jumps over the lazy dog
  # blahblaha
  masters = ["localhost:5050"]
  # The quick brown fox jumps over the lazy dog
  # blahblah
  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

**Comment by sparrc** [comment]
that's very odd, I wish I could get rid of the naoina toml parser but unfortunately I haven't been able to find anything that has the same level of functionality :(

**Comment by tripledes** [comment]
@sparrc so I you agree, I'll shorten the comments and push the change.

Regarding the TOML parser, didn't you guys wanted to fork it? I kind of remember some comments in the  influxdb project. I understand your point, but the naoina toml project seems rather dead...it hadn't got any activity for some months.

**Comment by tripledes** [comment]
@sparrc Another update...seems I'm like a bug triggering :bug: 

I know `memcached` plugin doesn't have all these fields, but was just trying to find what triggers it and whether it was only related to my plugin...so the following example makes the parser to panic again and the input mesos is completely gone from the config file.
But guess what... shortening the `master_collections` setting to something like `master_cols`, makes the parser happy again....so I guess it has been just a matter of luck that you didn't find this issue before.

```toml
# Read metrics from one or many memcached servers
[[inputs.memcached]]
  # Timeout, in ms.
  timeout = 100
  # The quick brown fox jumps over the lazy dog
  # blahblaha
  # The quick brown fox jumps over the lazy dog                                                                                                                                                               # blahblah                                                                                                                                                                                                  master_collections = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]                                                                                                 # An array of address to gather stats about. Specify an ip on hostname
  # with optional port. ie localhost, 10.0.0.1:11211, etc.
  servers = ["localhost:11211"]
  # unix_sockets = ["/var/run/memcached.sock"]
  # blabalsdhfla
  unix_sockets = ["resources","master","system","slaves","frameworks","messages","evqueue","registrar"]
```

I guess I'll shorten the setting to master_cols and leave the comments alone. Keep in mind this as I guess eventually someone else will hit it :)

**Comment by sparrc** [comment]
yes please push the change to shorten comments.....I'm not really following the status of our config library, but it certainly seems like naoina is not maintaining the library very actively.

**Comment by tripledes** [comment]
I know this makes no sense, but just moving all mesos settings to memcache plugin section and removing the mesos section, makes the parser to crash in the sam fashion....

Ok then, I'll push right away.

**Comment by sparrc** [comment]
:see_no_evil: :see_no_evil:

**Comment by sparrc** [comment]
thanks @tripledes, I'll try to get this merged tomorrow

File: internal/config/testdata/telegraf-agent.toml [hunk 3]
```diff
@@ -184,6 +184,18 @@
   # If no servers are specified, then localhost is used as the host.
   servers = ["localhost"]
 
+# Telegraf plugin for gathering metrics from N Mesos masters
+[[inputs.mesos]]
```
Comment (sparrc) [inline internal/config/testdata/telegraf-agent.toml]: can you make sure match the new SampleConfig?

File: plugins/inputs/mesos/README.md [hunk 4]
```diff
@@ -0,0 +1,168 @@
+# Mesos Input Plugin
+
+This input plugin gathers metrics from Mesos (*currently only Mesos masters*).
+For more information, please check the [Mesos Observability Metrics](http://mesos.apache.org/documentation/latest/monitoring/) page.
+
+### Configuration:
+
+```toml
+# Telegraf plugin for gathering metrics from N Mesos masters
+[[inputs.mesos]]
```
Comment (sparrc) [inline plugins/inputs/mesos/README.md]: can you make sure match the new SampleConfig?

**Comment by sparrc** [comment]
thanks @tripledes, just a couple places to make the new SampleConfig consistent

**Comment by sparrc** [comment]
merged, thanks @tripledes

**Comment by asdfsx** [comment]
excellent job!!! @tripledes

**Pull request closed without merging**